packages/expo-facebook/android/src/main/java/expo/modules/facebook/FacebookModule.kt

```diff
@@ -0,0 +1,339 @@
+package expo.modules.facebook
+
+import expo.modules.core.ExportedModule
+import com.facebook.CallbackManager
+import expo.modules.core.ModuleRegistry
+import com.facebook.appevents.AppEventsLogger
+import com.facebook.internal.AttributionIdentifiers
+import expo.modules.core.interfaces.ExpoMethod
+import com.facebook.FacebookSdk
+import com.facebook.AccessToken
+import android.os.Bundle
+import expo.modules.core.arguments.ReadableArguments
+import com.facebook.login.LoginManager
+import com.facebook.login.LoginBehavior
+import com.facebook.FacebookCallback
+import com.facebook.login.LoginResult
+import com.facebook.FacebookException
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import expo.modules.core.ModuleRegistryDelegate
+import expo.modules.core.Promise
+import expo.modules.core.interfaces.ActivityEventListener
+import expo.modules.core.interfaces.ActivityProvider
+import expo.modules.core.interfaces.services.UIManager
+import java.lang.Exception
+import java.lang.IllegalStateException
+import java.math.BigDecimal
+import java.util.*
+
+private const val ERR_FACEBOOK_MISCONFIGURED = "ERR_FACEBOOK_MISCONFIGURED"
+private const val ERR_FACEBOOK_LOGIN = "ERR_FACEBOOK_LOGIN"
+private const val PUSH_PAYLOAD_KEY = "fb_push_payload"
+private const val PUSH_PAYLOAD_CAMPAIGN_KEY = "campaign"
+
+class FacebookModule(
+  context: Context,
+  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
+) : ExportedModule(context), ActivityEventListener {
+  private val callbackManager: CallbackManager = CallbackManager.Factory.create()
+  private var appEventLogger: AppEventsLogger? = null
+  private var attributionIdentifiers: AttributionIdentifiers? = null
+  private var appId: String? = null
+  private var appName: String? = null
+  private val uIManager: UIManager by moduleRegistry()
+
+  private inline fun <reified T> moduleRegistry() = moduleRegistryDelegate.getFromModuleRegistry<T>()
+
+  override fun getName() = "ExponentFacebook"
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    moduleRegistryDelegate.onCreate(moduleRegistry)
+    uIManager.registerActivityEventListener(this)
+  }
+
+  override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
+    callbackManager.onActivityResult(requestCode, resultCode, data)
+  }
+
+  override fun onNewIntent(intent: Intent) = Unit
+
+  private fun bundleWithNullValuesAsStrings(parameters: ReadableArguments?): Bundle {
+    return Bundle().apply {
+      if (parameters != null) {
+        for (key in parameters.keys()) {
+          val value = parameters[key]
+          when (value) {
+            null -> putString(key, "null")
```

Comment:
```suggestion
      if (parameters != null) {
        for (entry in parameters) {
          when (entry.value) {
            null -> putString(entry.key, "null")
```
Up to you, IMO worth consideration